repos/iiif-storage/src/Utility/CheapOmekaRelationshipRequest.php

```diff
@@ -0,0 +1,141 @@
+SQL;
+
+    const SELECT_LABEL_RELATIONSHIP_WITH_LIMIT = <<<SQL
+SELECT V2.value_resource_id AS omeka_id, V1.uri AS uri
+ FROM value V2
+        LEFT JOIN value V1 on V1.resource_id = V2.value_resource_id
+ WHERE V2.property_id = :termId
+   AND V1.property_id = 10
+   AND V2.resource_id = :resourceId
+LIMIT :lim OFFSET :off;
+SQL;
+    const SELECT_URI_RELATIONSHIP_COUNT = <<<SQL
+SELECT COUNT(V2.value_resource_id) as count
+ FROM value V2
+        LEFT JOIN value V1 on V1.resource_id = V2.value_resource_id
+ WHERE V2.property_id = :termId
+   AND V1.property_id = 10
+   AND V2.resource_id = :resourceId;
+SQL;
+
+    const SELECT_SOURCE = <<<SQL
+SELECT value as json 
+  FROM value 
+  WHERE property_id = :termId 
+    AND resource_id = :resourceId
+SQL;
+
+    const ETAG = <<<SQL
+SELECT modified FROM resource WHERE id = :resourceId 
+SQL;
+
+
+
+    /**
+     * @param int $fromId
+     * @param string $relationTerm
+     * @param int $limit
+     * @param int $offset
+     * @return PaginatedResult
+     *
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    public function getUriMapping(int $fromId, string $relationTerm, int $limit = -1, int $offset = 0): PaginatedResult
+    {
+        $termId = $this->saturator->loadPropertyId($relationTerm);
+
+        $countStatement = $this->connection->prepare(self::SELECT_URI_RELATIONSHIP_COUNT);
+        $countStatement->bindValue('termId', (int)$termId, PDO::PARAM_INT);
+        $countStatement->bindValue('resourceId', (int)$fromId, PDO::PARAM_INT);
+        $countStatement->execute();
+
+        $statement = $this->connection->prepare($limit === -1 ? self::SELECT_URI_RELATIONSHIP : self::SELECT_LABEL_RELATIONSHIP_WITH_LIMIT);
+        $statement->bindValue('termId', (int)$termId, PDO::PARAM_INT);
+        $statement->bindValue('resourceId', (int)$fromId, PDO::PARAM_INT);
+        if ($limit !== -1) {
+            $statement->bindValue('lim', (int)$limit, PDO::PARAM_INT);
+            $statement->bindValue('off', (int)$offset, PDO::PARAM_INT);
+        }
+        $statement->execute();
+
+        $countResult = $countStatement->fetch();
+
+        $indexMap = [];
+        foreach ($statement->fetchAll() as $canvas) {
+            if ($canvas['uri']) {
+                $indexMap[$canvas['uri']] = $canvas['omeka_id'];
+            }
+        }
+
+        return new PaginatedResult(
+            $indexMap,
+            $countResult['count'] ?? 0,
+            $limit,
+            $offset
+        );
+    }
+
+    public function selectSource(int $fromId)
+    {
+        $dcTermsSource = $this->saturator->loadPropertyId('dcterms:source');
+        $sourceStatement = $this->connection->prepare(self::SELECT_SOURCE);
+        $sourceStatement->bindValue('termId', (int)$dcTermsSource, PDO::PARAM_INT);
+        $sourceStatement->bindValue('resourceId', (int)$fromId, PDO::PARAM_INT);
+        $sourceStatement->execute();
+        $sourceResult = $sourceStatement->fetch();
```

Comment:
Can we verify that the result set only contains 1 record here?

Reply:
Can't guarantee that, but its a computed field (the source JSON) so I've updated to close the cursor after getting the first result. It would take someone circumventing the CMS to add more than one JSON source though

Reply:
What I mean is: is it fine if we just pick the first result in a scenario where 2 come back, or would we want to say "this shouldn't have been possible"

Reply:
Yeh it's fine to pick just the first, if theres 2, someone has added that second field and should be ignored. Because the field is hidden from the UI by deafult, I don't think throwing an error here would be useful to an editor as they would not easily be able to fix the issue (or get in the situation in the first place)